Let each permission refine its algorithms and store data.

jyasskin opened the pull request:
This is now ready to review. See it in action at https://rawgit.com/jyasskin/permissions/allow-choosers/index.html, and a use of the new extension hooks at https://rawgit.com/jyasskin/web-bluetooth-1/permission-api/index.html#permission-api-integration.

domenic commented on index.html:
```diff
@@ -244,9 +269,71 @@
       <p>
-        The <code>PermissionName</code> enum defines the list of known
-        permission names.
+        Each enumerator-value in the <a>PermissionName</a> enum identifies a
```
enumerator-value is a weird term I've never heard before. Is it a C++ thing? Prefer "enumeration value" or "enum value".

jyasskin replied:
It's a funny mix of C++ and WebIDL. Oops. Fixing.

domenic commented on index.html:
```diff
@@ -244,9 +269,71 @@
-        The <code>PermissionName</code> enum defines the list of known
-        permission names.
+        Each enumerator-value in the <a>PermissionName</a> enum identifies a
+        <dfn data-export="">permission</dfn>, which consists of the following
```
Maybe it's against prevailing style in the file, but `<dfn data-export>` is equivalent and shorter.

jyasskin replied:
https://github.com/w3c/permissions/blob/gh-pages/contributing.md says to run tidy, which adds the `=""`. :-/

domenic commented on index.html:
```diff
@@ -244,9 +269,71 @@
       </p>
+      <dl>
+        <dt>
+          An <dfn data-export="">associated PermissionDescriptor</dfn>
```
Seems like things would be more consistent if this were "A permission descriptor type" to match the next two.

jyasskin replied:
I'd actually lean toward "permission query type", but "associated PermissionDescriptor" was already defined in the file, and that cleanup seemed like something I could plausibly leave to a subsequent PR. This one's already huge.

domenic commented on index.html:
```diff
@@ -244,9 +269,71 @@
+          A <dfn data-export="">permission query algorithm</dfn>
+        </dt>
+        <dd>
+          Takes an <a>associated PermissionDescriptor</a>, an instance of the
```
I think this would be "an instance of the associated permission descriptor type for this permission". The main thing is instance; right now associated PermissionDescriptor/associated permission descriptor type is a type, not an object.

jyasskin replied:
Done.

domenic commented on index.html:
```diff
@@ -244,9 +269,71 @@
+          <a>permission</a>, and an existing instance of the <a>permission
+          result type</a>, and updates the <a>permission result type</a>
+          instance with the query result. Used by <a for="Permissions" lt=
+          "query">Permissions.query()</a> and the <a href=
```
Probably also a prevailing style issue but this is pretty bad form; see https://github.com/whatwg/html/issues/826

jyasskin replied:
Done.

jyasskin commented:
I think this is ready for a serious look now, @marcoscaceres @mounirlamouri. Feel free to tell me to split it up into a couple different PRs.

domenic commented on index.html:
```diff
@@ -244,9 +269,71 @@
+          A <dfn data-export="">permission request algorithm</dfn>
+        </dt>
+        <dd>
+          Takes the previously-stored <a>permission storage type</a>, an
```
This paragraph needs a few "instance of"s inserted as well.

jyasskin replied:
Done.

domenic commented on index.html:
```diff
@@ -244,9 +269,71 @@
+        <dd>
+          Takes the previously-stored <a>permission storage type</a>, an
+          <a>associated PermissionDescriptor</a>, and a newly-created
+          <a>permission result type</a>, shows the user any necessary prompt to
```
s/shows/and shows

domenic replied:
Or maybe better s/, shows/. Shows

jyasskin replied:
Done.

domenic commented on index.html:
```diff
@@ -244,9 +269,71 @@
+          A <dfn data-export="">permission revocation algorithm</dfn>
+        </dt>
+        <dd>
+          After <a for="Permissions" lt="revoke">Permissions.revoke()</a>
```
This paragraph is inconsistent with the other two algorithms. Should be something like "Takes no arguments. Updates any other parts of the implementation that need to be kept in sync after an entry is removed from the permission store. Triggered by Permissions's revoke() method.

jyasskin replied:
Done.

domenic commented on index.html:
```diff
@@ -244,9 +269,71 @@
+          An <dfn data-export="">associated PermissionDescriptor</dfn>
+        </dt>
+        <dd>
+          A subtype of <a>PermissionDescriptor</a>.
```
Looking at all the examples, it seems like it's often not a subtype. Not sure how best to phrase it; maybe just a parenthatical (or PermissionDescriptor itself if a custom type is not needed).

jyasskin replied:
Hm, I didn't check the examples before writing this, but I see just:
```javascript
{name:'geolocation'}
{name:'notifications'}
{name:'camera', deviceId: devices[0].deviceId}
```
which are all instances of subtypes of  `PermissionDescriptor`:
```webidl
dictionary PermissionDescriptor {
    required PermissionName name;
};
```

What example are you looking at?

domenic replied:
PermissionDescriptor is not a subtype of PermissionDescriptor, at least not in any intuitive sense. I was looking at geolocations and notifications.

jyasskin replied:
Oh, I see. Strictly, types are subtypes of themselves, but not _proper_ subtypes of themselves, but I can word it differently. Maybe `<a>PermissionDescriptor</a> or one of its subtypes`?

domenic replied:
Right, outside of a mathematical context I'd err on the side of assuming subX = proper subX. That sounds pretty good to me.

jyasskin commented:
See https://rawgit.com/jyasskin/web-bluetooth-1/permission-api/index.html#permission-api-integration for the beginning of Web Bluetooth's use of these new hooks.

raymeskhoury commented:
Hey Jeffrey, a few random thoughts (some are not solely related to your edits):
-It initially hard to see which parts were external/exposed APIs vs internal constructs. e.g. PermissionDescriptor and PermissionStatus are external JS APIs - they must be complied with precisely whereas PermissionStorage is merely meant to help define storage behavior. Would it be helpful to draw a more clear line between those two groups?
-I also found it a bit hard to draw the distinction between required and recommended behaviors. It seems as though there should be room for UAs to handle permissions in ways that are outside of those currently specified behavior (for example, we have policy settings in Chrome that can override permissions). I don't think this is directly related to your edits.
-In relation to the above two points, perhaps a top-down restructuring of the spec could be helpful? External APIs and high level, required behaviors specified first and per-API behaviors, more subtle recommended behaviors, implementation details, etc. specified after? I'm no spec writer though :)
-I didn't think we were going to draw a distinction between an origin-based and embedding-based permission storage identifier instead leaving it up to the UA. I just pinged Mounir asking about this because we had chatted about it earlier. If it turns out to be the case does the permission storage identifier need to vary per-permission?

jyasskin commented:
Thanks for the comments!

> It initially hard to see which parts were external/exposed APIs vs internal constructs. e.g. PermissionDescriptor and PermissionStatus are external JS APIs - they must be complied with precisely whereas PermissionStorage is merely meant to help define storage behavior. Would it be helpful to draw a more clear line between those two groups?

Yeah. The traditional way to distinguish these would be to avoid describing `PermissionStorage` in IDL. However, that loses IDL's precision about what's stored, which I think is helping Bluetooth (I should be able to show you that tomorrow). I left a note saying `PermissionStorage` isn't exposed, but if you have other suggestions, I'm happy to take them.

>I also found it a bit hard to draw the distinction between required and recommended behaviors. It seems as though there should be room for UAs to handle permissions in ways that are outside of those currently specified behavior (for example, we have policy settings in Chrome that can override permissions). I don't think this is directly related to your edits.

Right, I think that's pre-existing. In most cases, I think we can say "the UA may customize the permission prompt, including auto-granting and auto-denying it" and "the UA may revoke at any time", and that's enough? I'd like to fix that in a separate PR, to keep this one from growing much more.

> In relation to the above two points, perhaps a top-down restructuring of the spec could be helpful? External APIs and high level, required behaviors specified first and per-API behaviors, more subtle recommended behaviors, implementation details, etc. specified after? I'm no spec writer though :)

Probably. Again, I want to keep rearrangements separate from the API extension.

> I didn't think we were going to draw a distinction between an origin-based and embedding-based permission storage identifier instead leaving it up to the UA. I just pinged Mounir asking about this because we had chatted about it earlier. If it turns out to be the case does the permission storage identifier need to vary per-permission?

Yeah, if it's always `PermissionName` + origin + UA-specific data, I think we can drop the identifier hook. I'll let Mounir confirm and then do that in my next update.

jyasskin commented:
The identifier hook is gone. PTAL.

mounirlamouri commented:
At a high level I agree with the changes here. Though, it's a very large CL and I can't realistically review this. Could you split this in a few parts:
- the commit that export the retrieve a permission (CC @tobie)
- the PermissionStorage definition
- the rest of the changes (I might ask you to split more but at least, I will have a better visibility -- keep the commits separated, it might be easier)

Also, could you remove tidy changes? It seems that we have issues with your tidy runs getting other results than mine. You are not the first person. I would like to figure this out before every one changes the spec to match its own tidy runs.

jyasskin commented:
Ok, I've split out a bunch of separate PRs that build on each other. I haven't yet undone the Tidy changes, and I've squashed the remaining changes in this PR since the individual commits were getting incoherent.

tobie commented:
> I haven't yet undone the Tidy changes

OK, sorry for bitching in the other PR, then. :/

jyasskin commented:
I've noticed that [PushMessagingPermissionContext::DecidePermission](https://code.google.com/p/chromium/codesearch/#chromium/src/chrome/browser/push_messaging/push_messaging_permission_context.cc&q=PushMessagingPermissionContext::DecidePermission&sq=package:chromium&type=cs&l=74) checks for `requesting_origin == embedding_origin`, which doesn't fit into the current [permission request algorithm](https://rawgit.com/jyasskin/permissions/allow-choosers/index.html#dfn-permission-request-algorithm) arguments.

If it's ok with you folks, I'd like to get the current PR merged, and then fix up details like this in subsequent smaller patches.
